### PR TITLE
[sonic_y_cable] add stub function for upgrade firmware of Y cable and split the get_part_number and get_vendor API's

### PR DESCRIPTION
**Pull request by vdahiya12**
This PR adds a stub function definition with details as to how it should be implemented by a Y cable vendor. 

#### Description
Added a stub function definition with a doc string describing how it should be implemented
Cleaning up and split some API's for vendors

#### Motivation and Context

Firmware upgrade is a functionality that vendors need to implement, this PR adds the definition 
and a description of how the implementation of this firmware_upgrade  API should be 

#### How Has This Been Tested?
Will be tested once implemented
for cleanup API's opened a python shell and executed API's for correctness

#### Additional Information (Optional)

Signed-off-by: vaibhav-dahiya <vdahiya@microsoft.com>

### CODE REVIEW

File: sonic_y_cable/y_cable.py [hunk 1]
```diff
@@ -1237,3 +1237,36 @@ def get_nic_voltage_temp(physical_port):
         return -1
 
     return temp, voltage
+
+def firmware_upgrade(physical_port, fwfile):
+
+    """ This routine should facilitate complete firmware
+    upgrade of the Y cable on all the three ends of the Y cable.
+    All the componenets of the Y cable should be upgraded
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1245]: s/componenets/components/

File: sonic_y_cable/y_cable.py [hunk 2]
```diff
@@ -1237,3 +1237,36 @@ def get_nic_voltage_temp(physical_port):
...
+    Args:
+        physical_port:
+             an Integer, the actual physical port connected to a Y cable
+        fwfile:
+             a File, the actual binary or executable which contains the firmware
+    Returns:
+        an Boolean:
+             True on sucessful firmware upgrade
+        an Integer:
+             an error code stating what was the cause of firmware upgrade failure
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1269]: Suggest keeping return value type consistent. If we want to handle multiple error codes, then we should define a success error code and some failures as constants, and ensure the function returns on of them.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: sonic_y_cable/y_cable.py (hunk 1 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 2 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py [hunk 3]
```diff
@@ -1237,3 +1243,37 @@ def get_nic_voltage_temp(physical_port):
         return -1
 
     return temp, voltage
+
+def firmware_upgrade(physical_port, fwfile):
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1247]: Suggest renaming to `upgrade_firmware`

File: sonic_y_cable/y_cable.py [hunk 4]
```diff
@@ -92,6 +92,12 @@
 
 MAX_NUM_LANES = 4
 
+# definitions of firmware upgrade codes
+# to be returned by the user when calling the
+# firmware upgrade routine
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:97]: Suggest simplifying to `# Valid return codes for upgrade_firmware() routine`

File: sonic_y_cable/y_cable.py [hunk 5]
```diff
@@ -1237,3 +1243,37 @@ def get_nic_voltage_temp(physical_port):
...
     return temp, voltage
+
+def firmware_upgrade(physical_port, fwfile):
+
+    """ This routine should facilitate complete firmware
+    upgrade of the Y cable on all the three ends of the
+    Y cable of the port specified.
+    All the components of the Y cable should be upgraded and committed
+    in their entirety by this single call subroutine.
+    This should return success code if firmware upgrade is successfull
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1254]: s/successfull/successful/

File: sonic_y_cable/y_cable.py [hunk 6]
```diff
@@ -1237,3 +1243,37 @@ def get_nic_voltage_temp(physical_port):
...
+    or synchronization would be required or not required after calling this
+    routine.
+    Meaning for example if one of the ends of the Y cable could not
+    complete the upgrade and the cable could then require synchronization
+    after this subroutine call, then an error code stating that should
+    be the return value. Likewise if all the links are down
+    then that would not require a rollback of any sort and hence the error code 
+    should just reflect that.
+    Hence after calling this subroutine the caller can then determine
+    whether further action is necessary for firmware upgrade or not.
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1266]: Suggest removing this section. Let's treat this function as an all-in-one firmware upgrade of all Y-cable ends that are possible from this end. If any part of the procedure fails, it should attempt to ensure the cable is in a functioning state and return an error code.

We can create more functions later if we want to allow for a more granular firmware upgrade (e.g., burn_firmware _image()`, `commit_new_firmware()`, etc.).

File: sonic_y_cable/y_cable.py [hunk 7]
```diff
@@ -1237,3 +1243,37 @@ def get_nic_voltage_temp(physical_port):
...
+
+    Args:
+        physical_port:
+             an Integer, the actual physical port connected to a Y cable
+        fwfile:
+             a File, the actual binary or executable which contains the firmware
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1272]: This should probably read `a String, a path to the binary file which contains the firmware image`

**Review by jleveque — CHANGES_REQUESTED** [review]

File: sonic_y_cable/y_cable.py (hunk 3 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 4 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 5 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 6 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 7 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

**Review by jleveque — APPROVED** [review]

**Review by jleveque — CHANGES_REQUESTED** [review]
Can you also split `get_nic_voltage_temp()` into two functions?

File: sonic_y_cable/y_cable.py [hunk 8]
```diff
@@ -999,7 +999,7 @@ def get_part_number(physical_port):
         if y_cable_validate_read_data(pn_result, 15, physical_port, "PN number") == EEPROM_READ_DATA_INVALID:
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:999]: s/PN number/part number/

File: sonic_y_cable/y_cable.py [hunk 9]
```diff
@@ -1024,7 +1024,7 @@ def get_vendor(physical_port):
         if y_cable_validate_read_data(result, 15, physical_port, "PN number") == EEPROM_READ_DATA_INVALID:
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1024]: s/PN number/part number/

**Review by jleveque — CHANGES_REQUESTED** [review]

File: sonic_y_cable/y_cable.py [hunk 10]
```diff
@@ -1237,3 +1262,90 @@ def get_nic_voltage_temp(physical_port):
         return -1
 
     return temp, voltage
+
+def get_internal_temperature(physical_port):
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1266]: Change from "internal" to "local" everywhere, as this is more descriptive that it is the temperature of the local connector, as opposed to the NIC or peer.

File: sonic_y_cable/y_cable.py [hunk 11]
```diff
@@ -1237,3 +1262,90 @@ def get_nic_voltage_temp(physical_port):
...
+        if y_cable_validate_read_data(result, 1, physical_port, "internal temperature") == EEPROM_READ_DATA_INVALID:
+            return EEPROM_ERROR
+        temp = result[0]
+    else:
+        helper_logger.log_error("platform_chassis is not loaded, failed to get internal temp")
+        return -1
+
+    return temp
+
+def get_internal_voltage(physical_port):
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1280]: Change from "internal" to "local" everywhere, as this is more descriptive that it is the temperature of the local connector, as opposed to the NIC or peer.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: sonic_y_cable/y_cable.py (hunk 10 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 11 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 8 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 9 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py [hunk 12]
```diff
@@ -1237,3 +1262,90 @@ def get_nic_voltage_temp(physical_port):
...
+        return -1
+
+    return voltage
+
+def get_nic_temperature(physical_port):
+
+    curr_offset = OFFSET_NIC_TEMPERATURE
+    if platform_chassis is not None:
+        result = platform_chassis.get_sfp(physical_port).read_eeprom(curr_offset, 1)
+        if y_cable_validate_read_data(result, 1, physical_port, "internal voltage") == EEPROM_READ_DATA_INVALID:
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1303]: s/internal voltage/NIC temperature/

File: sonic_y_cable/y_cable.py [hunk 13]
```diff
@@ -1237,3 +1262,90 @@ def get_nic_voltage_temp(physical_port):
...
+        return -1
+
+    return temp
+
+def get_local_voltage(physical_port):
+
+    if platform_chassis is not None:
+        curr_offset = OFFSET_INTERNAL_VOLTAGE
+        msb_result = platform_chassis.get_sfp(physical_port).read_eeprom(curr_offset, 1)
+        if y_cable_validate_read_data(msb_result, 1, physical_port, "local voltage msb") == EEPROM_READ_DATA_INVALID:
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1285]: s/msb/MSB/

File: sonic_y_cable/y_cable.py [hunk 14]
```diff
@@ -1237,3 +1262,90 @@ def get_nic_voltage_temp(physical_port):
...
+            return EEPROM_ERROR
+        lsb_result = platform_chassis.get_sfp(physical_port).read_eeprom(curr_offset+1, 1)
+        if y_cable_validate_read_data(lsb_result, 1, physical_port, "local voltage lsb") == EEPROM_READ_DATA_INVALID:
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1288]: s/lsb/LSB/

File: sonic_y_cable/y_cable.py [hunk 15]
```diff
@@ -1237,3 +1262,90 @@ def get_nic_voltage_temp(physical_port):
...
+        return -1
+
+    return temp
+
+def get_nic_voltage(physical_port):
+
+    curr_offset = OFFSET_NIC_VOLTAGE
+    if platform_chassis is not None:
+        msb_result = platform_chassis.get_sfp(physical_port).read_eeprom(curr_offset, 1)
+        if y_cable_validate_read_data(msb_result, 1, physical_port, "nic voltage msb") == EEPROM_READ_DATA_INVALID:
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1317]: s/nic/NIC/
s/msb/MSB/

File: sonic_y_cable/y_cable.py [hunk 16]
```diff
@@ -1237,3 +1262,90 @@ def get_nic_voltage_temp(physical_port):
...
+            return EEPROM_ERROR
+        lsb_result = platform_chassis.get_sfp(physical_port).read_eeprom(curr_offset+1, 1)
+        if y_cable_validate_read_data(lsb_result, 1, physical_port, "nic voltage lsb") == EEPROM_READ_DATA_INVALID:
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1320]: s/nic/NIC/
s/lsb/LSB/

**Review by jleveque — CHANGES_REQUESTED** [review]

File: sonic_y_cable/y_cable.py (hunk 12 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 13 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 14 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 15 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 16 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py [hunk 17]
```diff
@@ -977,34 +981,55 @@ def get_eye_info(physical_port, target):
...
+        if y_cable_validate_read_data(part_result, 15, physical_port, "Part number") == EEPROM_READ_DATA_INVALID:
             return EEPROM_ERROR
+    else:
+        helper_logger.log_error("platform_chassis is not loaded, failed to get part number")
+        return -1
+
+    part_number = str(part_result.decode())
+
+    return part_number
+
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1008]: Two blank lines between top-level functions per PEP8.

File: sonic_y_cable/y_cable.py [hunk 18]
```diff
@@ -977,34 +981,55 @@ def get_eye_info(physical_port, target):
...
-        helper_logger.log_error("platform_chassis is not loaded, failed to get pin results")
+        helper_logger.log_error("platform_chassis is not loaded, failed to get vendor")
         return -1
 
-    return pn_result, vendor_name
+    vendor_name = str(result.decode())
 
+    return vendor_name
 
 def get_switch_count(physical_port, count_type):
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1034]: Two blank lines between top-level functions per PEP8.

File: sonic_y_cable/y_cable.py [hunk 19]
```diff
@@ -1237,3 +1262,90 @@ def get_nic_voltage_temp(physical_port):
         return -1
 
     return temp, voltage
+
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1265]: Two blank lines between top-level functions per PEP8.

File: sonic_y_cable/y_cable.py [hunk 20]
```diff
@@ -1237,3 +1262,90 @@ def get_nic_voltage_temp(physical_port):
...
+        result = platform_chassis.get_sfp(physical_port).read_eeprom(curr_offset, 1)
+        if y_cable_validate_read_data(result, 1, physical_port, "local temperature") == EEPROM_READ_DATA_INVALID:
+            return EEPROM_ERROR
+        temp = result[0]
+    else:
+        helper_logger.log_error("platform_chassis is not loaded, failed to get local temp")
+        return -1
+
+    return temp
+
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1279]: Two blank lines between top-level functions per PEP8. Please do this throughout the file.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: sonic_y_cable/y_cable.py (hunk 20 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py:1282]: fixed

File: sonic_y_cable/y_cable.py (hunk 19 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py:1267]: fixed

File: sonic_y_cable/y_cable.py (hunk 18 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py:1036]: fixed

Comment (vdahiya12) [reply sonic_y_cable/y_cable.py:1036]: fixed

File: sonic_y_cable/y_cable.py (hunk 19 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py:1267]: fixed

File: sonic_y_cable/y_cable.py (hunk 20 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py:1282]: fixed

File: sonic_y_cable/y_cable.py (hunk 17 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py:1008]: fixed, used autopep8
autopep8 --max-line-length 120 -i y_cable.py

File: sonic_y_cable/y_cable.py [hunk 21]
```diff
@@ -977,33 +981,56 @@ def get_eye_info(physical_port, target):
...
+             an Integer, the actual physical port connected to a Y cable
+    Returns:
+        a string, with vendor name
+    """
 
+    curr_offset = OFFSET_VENDOR_NAME
+
+    if platform_chassis is not None:
+        result = platform_chassis.get_sfp(physical_port).read_eeprom(curr_offset, 15)
+        if y_cable_validate_read_data(result, 15, physical_port, "Part number") == EEPROM_READ_DATA_INVALID:
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1025]: s/Part number/Vendor name/

**Review by jleveque — CHANGES_REQUESTED** [review]

File: sonic_y_cable/y_cable.py (hunk 21 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

**Review by jleveque — APPROVED** [review]

**Pull request merged**